source/common/filter/tcp_proxy.cc

```diff
@@ -41,6 +46,13 @@ TcpProxyStats TcpProxyConfig::generateStats(const std::string& name, Stats::Stor
 }
 
 void TcpProxy::initializeUpstreamConnection() {
+  Upstream::ResourceManager& cluster_resource_manager =
+      cluster_manager_.get(config_->clusterName())
+          ->resourceManager(Upstream::ResourcePriority::Default);
+
+  if (!cluster_resource_manager.connections().canCreate()) {
+    return;
```

Comment:
I am closing downstream, and testing for that in the UpstreamConnectionLimit test.